node/lib/util/merge_util.js

```diff
@@ -129,294 +141,522 @@ exports.fastForwardMerge = co.wrap(function *(repo, commit, message) {
 });
 
 /**
- * Merge the specified `subs` in the specified `repo` having the specified
- * `index`.  Stage submodule commits in `metaRepo`.  Return an object
- * describing any commits that were generated and conflicted commits.  Use the
- * specified `opener` to acces submodule repos.  Use the specified `message` to
- * write commit messages.
- * @param {NodeGit.Repository} metaRepo
- * @param {Open.Opener}        opener
- * @param {NodeGit.Index}      metaIndex
- * @param {Object}             subs        map from name to SubmoduleChange
+ * Get the left side commit and sha for submodule `subRepo` based on workding
+ * directory type.
+ * 
+ * @async
+ * @param {NodeGit.Repository}  subRepo
+ * @param {SubmoduleChange}     subChange
+ * @param {SUB_OPEN_OPTION}     openOption
```

Comment:
Like, this isn't really about our opening policy -- it's about what we're merging.  There are two options: (a) our left side is our current state (working tree / index), vs (b) our left side is an arbitrary sha.

Reply:
correct. So basically the logic is now like this: 
1. `merge` command sets the `ourCommit` to null and `theirCommit` equal to a sha in the `MergeContext`. And `MergeContext` then set `ourCommit` to HEAD if current repo is not meata, and raises an error if it is.
2. `merge-bare` sets both `ourCommit` and `theirCommit`
3. left side commits of the merge for each submodule are also computed in the `MergeContext` based on the meta repo merging commits